### PR TITLE
[mlir][sparse] implement lowering rules for ExtractIterSpaceOp. 

### PR DESCRIPTION
**Pull request by PeimingLiu**
**DO NOT MERGE** until https://github.com/llvm/llvm-project/pull/89003

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.h [hunk 1]
```diff
@@ -288,10 +341,15 @@ std::unique_ptr<SparseTensorLevel> makeSparseTensorLevel(OpBuilder &builder,
...
 
-/// Helper function to create a simple SparseIterator object that iterate over
-/// the SparseTensorLevel.
-std::unique_ptr<SparseIterator> makeSimpleIterator(const SparseTensorLevel &stl,
-                                                   SparseEmitStrategy strategy);
+/// Helper function to create a TensorLevel object from given `tensor`.
+std::unique_ptr<SparseTensorLevel> makeSparseTensorLevel(LevelType lt, Value sz,
+                                                         ValueRange buffers,
+                                                         unsigned tid, Level l);
+/// Helper function to create a simple SparseIterator object that iterate
```
Comment (ingomueller-net) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.h:348]:
```suggestion
/// Helper function to create a simple SparseIterator object that iterates
```

Also: why has `over` been moved into the next line?

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.h [hunk 2]
```diff
@@ -80,6 +89,52 @@ enum class IterKind : uint8_t {
   kFilter,
 };
 
+class SparseIterationSpace {
```
Comment (ingomueller-net) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.h:92]: A bit of doc could be helpful here.

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.cpp [hunk 3]
```diff
@@ -94,8 +94,10 @@ class DenseLevel : public SparseTensorLevel {
 
   ValueRange getLvlBuffers() const override { return {}; }
 
-  ValuePair peekRangeAt(OpBuilder &b, Location l, ValueRange, Value p,
-                        Value max) const override {
+  ValuePair peekRangeAt(OpBuilder &b, Location l, ValueRange batchPrefix,
+                        ValueRange parentPos) const override {
+    assert(parentPos.size() == 1 && "Dense level can not be non-unique.");
```
Comment (ingomueller-net) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.cpp:99]:
```suggestion
    assert(parentPos.size() == 1 && "Dense level cannot be non-unique.");
```

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.cpp [hunk 4]
```diff
@@ -112,9 +114,9 @@ class BatchLevel : public SparseTensorLevel {
 
   ValueRange getLvlBuffers() const override { return {}; }
 
-  ValuePair peekRangeAt(OpBuilder &b, Location l, ValueRange, Value p,
-                        Value max) const override {
-    assert(max == nullptr && "Dense level can not be non-unique.");
+  ValuePair peekRangeAt(OpBuilder &b, Location l, ValueRange,
+                        ValueRange parentPos) const override {
+    assert(parentPos.size() == 1 && "Dense level can not be non-unique.");
```
Comment (ingomueller-net) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorIterator.cpp:119]:
```suggestion
    assert(parentPos.size() == 1 && "Dense level cannot be non-unique.");
```

**Review by ingomueller-net — COMMENTED** [review]
Looks good in general but I haven't understood all of the details. This particular op is probably good to start implementing because it is essentially a no-op, so you can concentrate on bringing all the glue code in place, but I am lacking some familiarity with the remainder of the code to really understand that.

File: mlir/test/Dialect/SparseTensor/sparse_space_collapse.mlir [hunk 5]
```diff
@@ -8,30 +8,31 @@
...
 // CHECK:         }
-func.func @sparse_sparse_collapse(%sp : tensor<4x8xf32, #COO>, %i : index) {
+func.func @sparse_sparse_collapse(%sp : tensor<4x8xf32, #COO>) -> index {
+  %i = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
   %l1 = sparse_tensor.extract_iteration_space %sp lvls = 0
       : tensor<4x8xf32, #COO>
      -> !sparse_tensor.iter_space<#COO, lvls = 0>
-  %r1 = sparse_tensor.iterate %it1 in %l1 at(%crd0) iter_args(%outer = %i): !sparse_tensor.iter_space<#COO, lvls = 0 to 1> -> index {
+    %r1 = sparse_tensor.iterate %it1 in %l1 iter_args(%outer = %i): !sparse_tensor.iter_space<#COO, lvls = 0 to 1> -> index {
```
Comment (yinying-lisa-li) [inline mlir/test/Dialect/SparseTensor/sparse_space_collapse.mlir:27]:
```suggestion
  %r1 = sparse_tensor.iterate %it1 in %l1 iter_args(%outer = %i): !sparse_tensor.iter_space<#COO, lvls = 0 to 1> -> index {
```

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseIterationToScf.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+using namespace mlir::sparse_tensor;
+
+void convertLevelType(SparseTensorEncodingAttr enc, Level lvl,
+                      SmallVectorImpl<Type> &fields) {
+  // Position and coordinate buffer in the sparse structure.
+  if (enc.getLvlType(lvl).isWithPosLT())
+    fields.push_back(enc.getPosMemRefType());
+  if (enc.getLvlType(lvl).isWithCrdLT())
+    fields.push_back(enc.getCrdMemRefType());
+  // One index for shape bound (result from lvlOp)
```
Comment (yinying-lisa-li) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseIterationToScf.cpp:20]:
```suggestion
  // One index for shape bound (result from lvlOp).
```

File: mlir/include/mlir/Dialect/SparseTensor/Transforms/Passes.td [hunk 7]
```diff
@@ -480,4 +480,19 @@ def SparseSpaceCollapse : Pass<"sparse-space-collapse", "func::FuncOp"> {
   ];
 }
 
+def LowerSparseIterationToSCF : Pass<"lower-sparse-iteration-to-scf", "func::FuncOp"> {
+  let summary = "lower sparse_tensor.iterate/coiterate into scf loops";
+  let description = [{
+     This pass lowers `sparse_tensor.iterate` operations into `scf.for/while` operations.
+     The pass is not yet stablized.
```
Comment (yinying-lisa-li) [inline mlir/include/mlir/Dialect/SparseTensor/Transforms/Passes.td:487]:
```suggestion
     The pass is not yet stabilized.
```

**Review by aartbik — APPROVED** [review]

**Pull request merged**